### PR TITLE
Introduce "kpack" Droplets and Support Custom Start Commands

### PR DESCRIPTION
**Pull request by tcdowney**
This PR introduces the concept of "kpack" droplets by associating the existing `kpack_lifecycle_data` for a build with its droplet instead of creating "docker" droplets. It allows us to distinguish kpack-built droplets from regular "docker" droplets and means we can now support custom start commands by invoking them with the CNB launcer: `/cnb/lifecycle/launcher`.

Additionally it includes a migration to convert existing kpack-built "docker" droplets into "kpack" droplets by associating their `kpack_lifecycle_data` with the "docker" droplet.

This change addresses the following issues:
* https://github.com/cloudfoundry/capi-k8s-release/issues/72
* https://github.com/cloudfoundry/cf-for-k8s/issues/502

CAKE Story: https://www.pivotaltracker.com/story/show/175081086

### CODE REVIEW

File: db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb
Comment (tcdowney) [inline db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb:1]: I hate making this sort of migration -- lots of potential for performance and data integrity issues -- so would appreciate a close look!

🕵️

File: db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb
Comment (tcdowney) [inline db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb:8]: I think there is an unlikely (?) race condition here where the `ccdb-migrate` job runs this but there are still `cf-api-servers` out there with the old code making "docker" droplets. If there are any apps in this state, they won't be able to get stack updates until the app is repushed.

I'm not sure if there is a way around that apart from either:
1. Reverting https://github.com/cloudfoundry/cloud_controller_ng/pull/1917/commits/91769482589608ed0ae664da9e8b33b70c4f29ff and always supporting the ability to update image for "docker" and "kpack" droplets.
2. Running the migration again as part of a future release.

File: lib/cloud_controller/opi/apps_client.rb
Comment (jspawar) [inline lib/cloud_controller/opi/apps_client.rb:121]: I suspect this was intentional and fine, but for my knowledge, why does this remain `docker_lifecycle` instead of something new like `kpack_lifecycle`? Is this because Eirini itself is looking for specific information under that key when it desires a LRP?

File: lib/cloud_controller/opi/apps_client.rb
Comment (tcdowney) [inline lib/cloud_controller/opi/apps_client.rb:116]: I think this command should actually be:
```ruby
['/bin/sh', '-c', "#{CNB_LAUNCHER_PATH} #{@process.command}"]
```

where the launcher and specified command are a single string that's passed to the shell.

The app I had tested this with was working, but I think it may have been running the images default entrypoint? I don't know... a bit confused by that. Anyways going to add a commit for this.

Comment (tcdowney) [reply lib/cloud_controller/opi/apps_client.rb]: https://github.com/cloudfoundry/cloud_controller_ng/pull/1917/commits/6a32abbd93ca77604f2657347cdb6522b647f857

File: lib/cloud_controller/opi/apps_client.rb
Comment (tcdowney) [inline lib/cloud_controller/opi/apps_client.rb:115]: Does anyone know why we use `@process.command` here instead of `@process.started_command`?

https://github.com/cloudfoundry/cloud_controller_ng/blob/bd3d96ca9a959936c6f655049e155b3edc7a7c8b/lib/cloud_controller/diego/buildpack/lifecycle_protocol.rb#L74

I'm having trouble understanding how this picks of start commands for other processes that are defined in the `Procfile`.

File: lib/cloud_controller/opi/apps_client.rb
Comment (tcdowney) [reply lib/cloud_controller/opi/apps_client.rb:121]: Yeah, exactly.

Eirini is looking for either a `DockerLifecycle` or a `BuildpackLifecycle` (which I guess is only used in `kube-cf`):
https://github.com/cloudfoundry-incubator/eirini/blob/ee87b6612c6832f26a42a55ee309c222eee36bcb/bifrost/convert.go#L140

File: db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb
Comment (tcdowney) [reply db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb:9]: Yeah I think that's the same as what I mentioned in this comment:
https://github.com/cloudfoundry/cloud_controller_ng/pull/1917#discussion_r508857218

I'm not sure if there is a way around it apart from either:

* Reverting 9176948 and always supporting the ability to update image for "docker" and "kpack" droplets.
* Running the migration again as part of a future release (this requires people to upgrade to this version first -- if they jump and skip this version then same problem)
* A Kubernetes job that runs on every cf-for-k8s upgrade and migrates them which we eventually delete? 🤔
* We accept it as an edge case and document the workaround (repush) if stack updates fail for an app

File: db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb
Comment (tcdowney) [reply db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb:8]: Resolving this, let's move discussion to this thread:
https://github.com/cloudfoundry/cloud_controller_ng/pull/1917#discussion_r508889764

File: db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb
Comment (cwlbraa) [reply db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb:9]: > I think there is an unlikely (?) race condition here where the ccdb-migrate job runs this but there are still cf-api-servers out there with the old code making "docker" droplets. If there are any apps in this state, they won't be able to get stack updates until the app is repushed (repushing is a fairly easy mitigation IMO).

> Scenario I'm imagining would require one cf-api-server pod to rollout with new changes and one to remain with old changes. cf push gets its request routed to the old one => droplet eventually created without associated kpack_lifecycle_data

This is almost certainly possible, and was not something I had considered when Tim and I talked about this earlier. One thing that we've never been able to build on BOSH that would totally fix this type of deploy scenario is a post-deploy migration job in addition to the pre-deploy one. If we had that running its own, idempotent post-deploy migrations, we could clean up after ourselves in scenarios like this and maybe even use it implement proper [expand-and-contract parallel-change migrations](https://martinfowler.com/articles/evodb.html).

> We accept it as an edge case and document the workaround (repush) if stack updates fail for an app

I think we're still early enough in the life of cf-for-k8s that this is acceptable, particularly if you consider this relative to how broken other things are.

File: lib/cloud_controller/opi/apps_client.rb
Comment (tcdowney) [reply lib/cloud_controller/opi/apps_client.rb:115]: Heh when I switch to `process.started_command` it doesn't work cause we run into a new bug. I don't think our build reconciler is copying over detected start commands correctly.

https://github.com/cloudfoundry/capi-k8s-release/issues/88

File: db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb
Comment (tcdowney) [reply db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb:9]: > I think we're still early enough in the life of cf-for-k8s that this is acceptable, particularly if you consider this relative to how broken other things are

I think it may be cheaper to just leave this in now versus doing it later and risking the context being lost. 🤷 

Cause of issues like https://github.com/cloudfoundry/capi-k8s-release/issues/88 where we're literally missing data on some of the droplets, I think folks may have to repush anyway.

Comment (tcdowney) [reply db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb:9]: Actually, as I think about this more... we might be able to use the fact that historical kpack-built droplets are "docker" to help resolve that issue (https://github.com/cloudfoundry/capi-k8s-release/issues/88).

It will be difficult to know when it is safe for a "kpack" droplet to use `@process.started_command` if some "kpack" droplets have the incomplete command, but if we fix that issue at the same time as we introduce "kpack" droplets then we can assume that all of them have complete detected start commands. 🤔

Comment (cwlbraa) [reply db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb:9]: > It will be difficult to know when it is safe for a "kpack" droplet to use @process.started_command if some "kpack" droplets have the incomplete command, but if we fix that issue at the same time as we introduce "kpack" droplets then we can assume that all of them have complete detected start commands. 🤔

This is a good point. It's a kinda risky two-birds-with-one-stone way of fixing these bugs, but it sounds like it'd work.

Effectively what you're saying is _dont_ turn any docker droplets into kpack droplets. Instead, require repushing. 

This way, we'll know that kpack droplets have usable start commands and the migration won't break any apps that just happen to work in the current, weird state. Otherwise, switching to use `started_command` could break apps on migration because we weren't using them before and were likely saving the wrong data.

File: lib/cloud_controller/opi/apps_client.rb
Comment (jspawar) [reply lib/cloud_controller/opi/apps_client.rb:121]: Should we teach eirini to speak `KpackLifecycle`? Can merge this and revisit after if we really want to

File: db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb
Comment (tcdowney) [reply db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb:9]: > Otherwise, switching to use started_command could break apps on migration because we weren't using them before and were likely saving the wrong data.

Yep, when I tried doing this in this branch `dora` would just run `bundle` and install gems instead of `bundle rackup`.

File: lib/cloud_controller/opi/apps_client.rb
Comment (tcdowney) [reply lib/cloud_controller/opi/apps_client.rb:121]: IMO we shouldn't until there is a reason to. I feel like at this point if Eirini doesn't need to distinguish between an app using a "regular" Docker image and an app using a kpack-built Docker image it shouldn't have to.

File: lib/cloud_controller/opi/apps_client.rb
Comment (tcdowney) [reply lib/cloud_controller/opi/apps_client.rb:115]: I put out a PR to fix the detected start commands we're saving on droplets:
https://github.com/cloudfoundry/capi-k8s-release/pull/90

Once that is merged I can revert the "docker" -> "kpack" migration and switch this over to using `@process.started_command`.

**Review by cwlbraa — CHANGES_REQUESTED** [review]
This looks good, but given what we've learned about the current state of PATCH droplet.process_types, the migration to turn docker droplets into kpack droplets and then use those seems likely to break apps that just happen to work currently.

File: spec/unit/actions/droplet_create_spec.rb
Comment (tcdowney) [reply spec/unit/actions/droplet_create_spec.rb:269]: existing behavior 👻

File: app/actions/droplet_create.rb
Comment (tcdowney) [reply app/actions/droplet_create.rb:74]: > assign the lifecycle data... why does this happen after save?

The droplet in this case hasn't been persisted yet (and I think doesn't have a `guid`) until it's saved. It's kinda tricky, but `droplet.kpack_lifecycle_data = build.kpack_lifecycle_data` actually sets the `kpack_lifecycle_data.droplet_guid` field in a different table.

The `droplet.reload` I assume is for the events... I actually just saw that the regular buildpack droplet create was doing this and figured it was important for that.

File: db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb
Comment (tcdowney) [reply db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb]: I rebased on master and removed the migration as part of that. It's no longer present in this commit:
https://github.com/cloudfoundry/cloud_controller_ng/pull/1917/commits/87d0369f72ccd411b6ce2551d42105b1d9e6a48a

File: db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb
Comment (tcdowney) [reply db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb]: This migration is now gone. Resolving.

File: lib/cloud_controller/opi/apps_client.rb
Comment (tcdowney) [reply lib/cloud_controller/opi/apps_client.rb]: That's merged, so I switched this over:
https://github.com/cloudfoundry/cloud_controller_ng/pull/1917/commits/0a55501093d6a10472274f8bc134dbf4bbe60f9f

**Comment by tcdowney** [comment]
The unit test failures are unrelated. It looks like these have been failing on Github Actions on other PRs (with unrelated changes) as well. :/
E.g. https://github.com/cloudfoundry/cloud_controller_ng/pull/1920/checks?check_run_id=1287920643

File: db/migrations/20201020204056_index_kpack_lifecycle_data_droplet_guid.rb
Comment (cwlbraa) [inline db/migrations/20201020204056_index_kpack_lifecycle_data_droplet_guid.rb:4]: This strikes me as an obvious improvement, but I've never thought about indices on `lifecycle_data` tables prior to now. It seems likely that we've missed some other indices on this table in the past `lifecycle_data` links itself to all sorts of other objects...

File: lib/cloud_controller/opi/apps_client.rb
Comment (cwlbraa) [inline lib/cloud_controller/opi/apps_client.rb:116]: 🤔  is the `bin/sh` necessary here? it's safe to leave it in but it's kinda interesting to think about.

File: app/actions/droplet_create.rb
Comment (cwlbraa) [reply app/actions/droplet_create.rb:74]: > droplet.kpack_lifecycle_data = build.kpack_lifecycle_data actually sets the kpack_lifecycle_data.droplet_guid field in a different table.

lol this is definitely why people don't like ORMs

**Review by cwlbraa — APPROVED** [review]
lgtm

File: db/migrations/20201020204056_index_kpack_lifecycle_data_droplet_guid.rb
Comment (tcdowney) [reply db/migrations/20201020204056_index_kpack_lifecycle_data_droplet_guid.rb:4]: I put a little blurb in the commit message about this one:

```
- add index to kpack_lifecycle_data.droplet_guid
  - this is important since droplets find their associated lifecycle
  through this column
  - a similar index exists on buildpack_lifecycle_data
```

I think it's important if we want the `droplet.kpack?` type of methods to be performant.

File: lib/cloud_controller/opi/apps_client.rb
Comment (tcdowney) [reply lib/cloud_controller/opi/apps_client.rb:116]: Yeah, I thought that as well. I went with this cause it's what we were doing for docker droplets, but 🤷 .
https://github.com/cloudfoundry/cloud_controller_ng/pull/1417/files#diff-738bfbb749a5827ff834ca07b3594487e3f321066cf64d20ff4bdf7359137f4bR74

I can ask the Eirini folks.

**Review by jspawar — APPROVED** [review]
ya lgtm

**Pull request merged**
